File: 0812/SeoyeonHong/연속된 부분 수열의 합.py
# https://school.programmers.co.kr/learn/courses/30/lessons/178870
def solution(sequence, k):
    l = len(sequence)
    answer = [0, l-1]
    min_len = l
    
    start, end = 0, 0
    sub_sum = sequence[0]
    while start <= end < l:
        if sub_sum < k and end < l-1: # 부분 수열의 합이 k보다 작고 end를 증가시킬 수 있다면
            end += 1
            sub_sum += sequence[end]
        elif sub_sum < k: # 부분 수열의 합이 k보다 작고 end가 l-1이라면 종료
            break
        elif sub_sum >= k: # 부분 수열의 합이 k이상이면 start + 1
            if sub_sum == k and min_len > end - start + 1: # 부분수열의 합이 k이면
                min_len = end - start + 1 # 길이가 짧은 수열 갱신
                answer = [start, end] 
            sub_sum -= sequence[start]
            start += 1
            if end < start < l:
                end += 1
                sub_sum += sequence[end]
                            
    return answer

File: 0812/AyoungKim/연속된 부분 수열의 합.java
// 합이 k가 되는 부분 수열의 시작인덱스와 끝 인덱스를 구하는 문제
// 합이 k가 되는 부분 수열이 많을 경우, 길이가 짧은 수열을 구함
// 길이가 짧은 수열이 여러 개일 경우, 시작 인덱스가 작은 수열을 구함

class Solution {
    public int[] solution(int[] sequence, int k) {
        int start = 0;
        int end = 0;
        int min_len = Integer.MAX_VALUE;
        int sum = sequence[0];

        int n = sequence.length;
        int cur_start = 0;
        int cur_end = 0;

        while (cur_start < n && cur_end < n) {

            // 만약 합이 k가 되는 부분 수열을 찾았을 경우
            if (sum == k) {
                // 수열의 길이를 비교

                // 만약 현재 저장된 길이보다 작다면
                if (min_len > cur_end - cur_start) {
                    min_len = cur_end - cur_start;
                    start = cur_start;
                    end = cur_end;
                } else if (min_len == cur_end - cur_start) {
                    // 만약 길이가 같다면 시작 인덱스 값 비교
                    if (start > cur_start) {
                        start = cur_start;
                        end = cur_end;
                    }
                }
                sum -= sequence[cur_start++];
            } else if (sum < k) {
                // 합이 k보다 작을 경우 끝 인덱스를 늘림
                if (cur_end == n - 1) break;
                sum += sequence[++cur_end];
            } else {
                // 합이 k보다 클 경우 시작 인덱스를 땡김
                sum -= sequence[cur_start++];
            }
        }

        int[] answer = {start, end};
        return answer;
    }
}